### PR TITLE
Mpfa cleanup, and vector sources for finite volume elliptic methods

### PR DESCRIPTION
**Pull request by keileg**
## Main changes:
* The internal logic in the mpfa discretization is reworked, based on logic from mpsa / biot. This significantly simplified the implementation.
* Mpfa can now be constructed gradually (reduced memory need), or partially updated.
* Mpfa and tpfa discretize and assemble vector source terms (for instance gravity).

### CODE REVIEW

File: src/porepy/numerics/fv/tpfa.py [hunk 1]
```diff
@@ -69,26 +69,19 @@ def discretize(self, g, data, vector_source=False, faces=None):
         ----------
         g (pp.Grid): grid, or a subclass, with geometry fields computed.
         data (dict): For entries, see above.
-        vector_source (Boolean): optional. Defines flux contribution from vector source
-        faces (np.ndarray): optional. Defines active faces.
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/tpfa.py:73]: Do we want to provide a summary of the keywords of data (or rather of data["parameters"]) which are accessed by the discretize method? I.e. "second_order_tensor" etc

File: test/integration/test_2d_benchmark_geiger.py [hunk 2]
```diff
@@ -119,4 +119,5 @@ def test_mpfa_permeable(self):
 
 
 if __name__ == "__main__":
+    TestFVOnBenchmark().test_tpfa_blocking()
```
Comment (IvarStefansson) [inline test/integration/test_2d_benchmark_geiger.py:122]:
```suggestion
```

File: test/unit/test_darcy_mortar.py [hunk 3]
```diff
@@ -1084,4 +1087,5 @@ def test_RT0_refined_1d(self):
 
 
 if __name__ == "__main__":
+    TestMortar2dSingleFractureCartesianGrid().test_tpfa_matching_grids_uniform_flow_larger_domain()
```
Comment (IvarStefansson) [inline test/unit/test_darcy_mortar.py]:
```suggestion
```

File: test/unit/test_mpfa_gravity.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import numpy as np
 import scipy
+from scipy.sparse.linalg import spsolve
```
Comment (IvarStefansson) [inline test/unit/test_mpfa_gravity.py:3]: I don't see where this is used below
```suggestion
```

File: src/porepy/numerics/fv/mpfa.py [hunk 5]
```diff
@@ -52,14 +51,12 @@ def discretize(self, g, data):
                 the permeability (such as with fracture apertures) should be
                 included in the permeability.
             bc: (BoundaryCondition) boundary conditions
-            vector_source: (boolean) Optional. Discretization of the divergence of
-                a vector source. For example, gravity in the flow equations. If False,
-                standard MPFA is used.
+            vector_source_dim: (int) Optional. Ambient dimension, used in the
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:54]: I didn't see this in tpfa

File: src/porepy/numerics/fv/mpfa.py [hunk 6]
```diff
@@ -72,461 +69,260 @@ def discretize(self, g, data):
...
-            rhs = q + div * bound_flux * bound_vals
-            x = sps.linalg.spsolve(A, rhs)
-            # Recover flux
-            f = flux * x - bound_flux * bound_vals
-            # Recover boundary pressure
-            bp = bp_cell * x + bp_face * bound_vals
-        """
+        # NOTE: active_faces are all faces to have their stencils updated, while
+        # active_cells may form a larger set (to accurately update all faces on a
+        # subgrid, it is necessary to assign some overlap in terms cells).
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:100]: I'm not entirely sure what "in terms of" means here, but I think there should be an "of"
```suggestion
        # subgrid, it is necessary to assign some overlap in terms of cells).
```

File: src/porepy/numerics/fv/mpfa.py [hunk 7]
```diff
@@ -72,461 +69,260 @@ def discretize(self, g, data):
...
+            # Split the discretization.
+            (
+                loc_flux,
+                loc_bound_flux,
+                loc_bound_pressure_cell,
+                loc_bound_pressure_face,
+                loc_vector_source,
+            ) = discr_fields
+
+            # Next, transfer discretization matrices from the local to the active grid
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:177]: What is "local" in this context?

File: src/porepy/numerics/fv/mpfa.py [hunk 8]
```diff
@@ -72,461 +69,260 @@ def discretize(self, g, data):
...
+        if g.dim == 2:
+            # Use the same mapping of the geometry as was done in
+            # self._flux_discretization(). This mapping is deterministic, thus the
+            # rotation matrix should be the same as applied before. In this case, we
+            # only need the rotation, and the active dimensions
+            *_, R, dim, _ = pp.map_geometry.map_grid(g)
+
+            # We need to pick out the parts of the rotation matrix that gives the
+            # in-plane (relative to the grid) parts, and apply this to all cells in the
+            # grid. The simplest way to do this is to expand R[dim] via sps.block_diags,
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:236]: Where is the normal part handled?

File: src/porepy/numerics/fv/mpfa.py [hunk 9]
```diff
@@ -1379,7 +1141,50 @@ def _create_bound_rhs(
...
+        whether the implementation is sufficiently general to be put there.
+
+        Parameters:
+            sub_g (pp.Grid): Grid for which the new condition applies. Is
+                assumed to be a subgrid of the grid to initialize this object.
+            face_map (np.ndarray): Index of faces of the original grid from
+                which the new conditions should be picked.
+
+        Returns:
+            BoundaryConditionVectorial: New bc object, aimed at a smaller grid.
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:1162]: Vectorial?

File: src/porepy/numerics/fv/mpfa.py [hunk 10]
```diff
@@ -1379,7 +1141,50 @@ def _create_bound_rhs(
...
+        return sub_bc
+
+    def _constit_for_subgrid(
+        self, constit: pp.SecondOrderTensor, loc_cells: np.ndarray
+    ) -> pp.SecondOrderTensor:
+        # Copy stiffness tensor, and restrict to local cells
+        loc_c = constit.copy()
+        loc_c.values = loc_c.values[::, ::, loc_cells]
+        # Also restrict the lambda and mu fields; we will copy the stiffness
+        # tensors later.
```
Comment (IvarStefansson) [inline src/porepy/numerics/fv/mpfa.py:1186]: I don't understand the comment. Does this mean that the function is yet to be expanded to fourth order tensors? TODO?

File: src/porepy/grids/grid.py [hunk 11]
```diff
@@ -116,7 +116,7 @@ def __init__(
 
         # Add tag for the boundary faces
         if tags is None:
-            self.tags: Dict = {}
+            self.tags: Dict[str, np.ndarray] = {}
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/grids/grid.py:119]: Yes, that is the guideline. This is what we do, is it not?

We could of course be more specific, and say np.ndarray[bool] or something, but I don't think that works as intended - last time I checked, typing and numpy was not exactly a perfect match.

File: src/porepy/numerics/fv/fvutils.py [hunk 12]
```diff
@@ -244,9 +244,103 @@ def determine_eta(g):
...
+    # Find the cells and faces that should be considered for discretization
+    if (
+        (specified_cells is not None)
+        or (specified_faces is not None)
+        or (specified_nodes is not None)
+    ):
+        # Find computational stencil, based on specified cells, faces and nodes.
+        active_cells, active_faces = pp.fvutils.cell_ind_for_partial_update(
+            g, cells=specified_cells, faces=specified_faces, nodes=specified_nodes
+        )
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/fvutils.py:284]: No, it should be fine to combine them.

File: src/porepy/numerics/fv/fvutils.py [hunk 13]
```diff
@@ -244,9 +244,103 @@ def determine_eta(g):
...
+
+        # Extract subgrid, together with mappings between local and active
+        # (global, or at least less local) cells
+        sub_g, l2g_faces, _ = pp.partition.extract_subgrid(active_grid, loc_cells)
+        l2g_cells = sub_g.parent_cell_ind
+
+        yield sub_g, loc_faces, cells_in_partition, l2g_cells, l2g_faces
+
+
+def remove_nonlocal_contribution(raw_ind: np.ndarray, nd: int, *args: Any) -> None:
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/fvutils.py:337]: Thanks. Will do.

File: src/porepy/numerics/fv/fvutils.py [hunk 14]
```diff
@@ -1319,6 +1417,7 @@ def map_subgrid_to_grid(
             corresponding cell in the larger grid.
         is_vector (bool): If True, the returned mappings are sized to fit with vector
             variables, with g.dim elements per cell and face.
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/fvutils.py:1419]: Indeed.

File: src/porepy/numerics/fv/mpfa.py (hunk 5 above)
Comment (keileg) [reply src/porepy/numerics/fv/mpfa.py:54]: You are right. I added the tpfa functionality towards the end, but have not fixed embedded grids. That also shows the test coverage is not good enough.

File: src/porepy/numerics/fv/mpfa.py (hunk 7 above)
Comment (keileg) [reply src/porepy/numerics/fv/mpfa.py:177]: Refers to the subproblems (splitting due to max_memory).

File: src/porepy/numerics/fv/mpfa.py (hunk 8 above)
Comment (keileg) [reply src/porepy/numerics/fv/mpfa.py:236]: Not sure - perhaps it enters the interface law? Something for you, I am afraid.

File: src/porepy/numerics/fv/mpfa.py (hunk 9 above)
Comment (keileg) [reply src/porepy/numerics/fv/mpfa.py:1162]: oops.

File: src/porepy/numerics/fv/mpfa.py (hunk 10 above)
Comment (keileg) [reply src/porepy/numerics/fv/mpfa.py:1186]: Leftover from copy from the similar funciton in mpsa. My bad.

File: src/porepy/numerics/fv/tpfa.py (hunk 1 above)
Comment (keileg) [reply src/porepy/numerics/fv/tpfa.py:73]: It is already there, in the discretize() method.

**Pull request merged**